index.js

```diff
@@ -62,15 +61,33 @@ const useNavigate = (options) => {
  * Part 2, Low Carb Router API: Router, Route, Link, Switch
  */
 
-export const Router = (props) => {
-  // this little trick allows to avoid having unnecessary
-  // calls to potentially expensive `buildRouter` method.
+export const Router = ({
+  hook,
+  matcher,
+  base = "",
+  nested = false,
+  children,
+}) => {
+  const parent = useRouter();
+
+  // nested `<Router />` has the scope of its closest parent router (base path is prepended)
+  // Routers are not nested by default, but this might change in future versions
+  const proto = nested ? parent : defaultRouter;
+
+  // this component doesn't handle prop updates, it is done intentionally to avoid unexpected
+  // side effects (e.g. you can't just hot swap the hook, it'll break the routing)
+  //
+  // we use `useState` here, but it only catches the first render and never changes.
   // https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily
-  const [value] = useState(() => ({ v: buildRouter(props) }));
+  const [value] = useState(() => ({
+    hook: hook || proto.hook,
+    matcher: matcher || proto.matcher,
+    base: proto.base + base,
+  }));
```

Comment:
Looks good, except we **_absolutely do_** still need a layout effect here to update props (for **AT LEAST** `base`, preferably `matcher` as well).
Consider the following JSX:

```typescript
return app1Flag ? 
    <Router base="/app1">{children}</Router> 
    : 
    <Router base="/app2">{children}</Router>;
```
This code will fail dramatically if `base` is not updated within your ref (because React will simply change the props of the current component). My own use-cases would not work so I would still have to resort to hacks. In short: _we cannot assume that routers are immutable_. Now, one thing we can do is assume that the `hook` is immutable and simply not update the hook (a most-likely valid assumption). But no need to throw the baby out with the bathwater, because we can expect `base` to change frequently, depending on what part the app is currently being rendered.

Here's my recommended addition:

```typescript
useLayoutEffect(() => {
    value.matcher = matcher || proto.matcher;
    value.base = proto.base + base;
});
```

In addition, it would certainly be nice to have a reference to the parent router (e.g., if we want to navigate from the context of the _parent's_ base path in some custom operation.)

So I would further recommend the following change:

```typescript
const [value] = useState(() => ({
  hook: hook || proto.hook,
  matcher: matcher || proto.matcher,
  base: proto.base + base,
  parent: proto // <-- inserted here (note: `parent` for the default router should still be undefined, so we can detect when we are at the top).
}));
useLayoutEffect(() => {
    value.matcher = matcher || proto.matcher;
    value.base = proto.base + base;
    value.parent = proto;
});
```

Reply:
You have a point here, I'll update this part.